### PR TITLE
[export] Refactor `constrain_as_value` and `constrain_as_size`

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106591

Some notable changes:
1. `constrain_as_size` allows min value to be less than 2 as it will unconditionally assume min >= 2 for compiler purposes. Instead, we add additional check to make sure max value is always greater than 2. 
2. Previously, we used to runtime assert on the unbacked symint's val range which would be always between [2, max]. I modified this logic to assert on [0, max] unless user explicitly specifies the min range. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @anijain2305

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -183,12 +183,11 @@
 
 - func: sym_constrain_range(Scalar size, int? min=None, int? max=None) -> ()
   dispatch:
-    CPU: sym_constrain_range_cpu
-    CUDA: sym_constrain_range_cuda
+    CompositeExplicitAutograd: sym_constrain_range
```
[the remark this answers is not in the document]
Comment (tugsbayasgalan) [reply aten/src/ATen/native/native_functions.yaml:186]: I think so lol

File: test/dynamo/test_export.py [hunk 2]
```diff
@@ -2321,7 +2321,7 @@ def test_exported_graph_serialization(self):
 
         def f(x, y):
             b = x.item()
-            constrain_as_size(b, min=2, max=5)
+            constrain_as_size(b, max=5)
```
Comment (ezyang) [inline test/dynamo/test_export.py:2324]: oh, you didn't end up deleting min/max from constrain_as_size?

File: torch/fx/experimental/symbolic_shapes.py [hunk 3]
```diff
@@ -346,13 +367,9 @@ def constrain_range(a, *, min: Optional[int], max: Optional[int] = None):
         values at runtime (we assume that a graph that is valid for N=2 will
         also be valid for N=1).
     """
+    assert max is not None
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:370]: How come? 🤔

File: torch/_export/constraints.py [hunk 4]
```diff
@@ -1,48 +1,96 @@
...
+def constrain_as_size(symbol, max: Optional[int] = None):
     """
-    Add min/max constraint on the intermediate symbol which will be used as a size
+    Add max constraint on the intermediate symbol which will be used as a size
+    in another tensor. If called in eager mode, it will still check if the input
+    value is within the specified range.
     """
+    # NOTE: If min value is not passed, we will assume it means there is no
+    # user specified min value provided. If this happens, we will assume min value to
+    # be 2 for compiler only! Runtime will assume min value will be 0.
```
Comment (ezyang) [inline torch/_export/constraints.py:81]: But there's no int anymore

File: torch/_export/constraints.py [hunk 5]
```diff
@@ -1,48 +1,96 @@
...
 
-    # TODO: we should investigate turning off 0/1 specialization for unbacked
-    # SymInts
-    if min < 2:
-        raise ValueRangeError(
-            "Unable to set min size to be <= 2 because we specialize on 0/1 sizes."
-        )
-    return constrain_as_value(symbol, min, max)
+    assert isinstance(symbol, SymInt)
+    return torch.sym_constrain_range(symbol, min=None, max=max)
```
Comment (ezyang) [inline torch/_export/constraints.py:96]: Doesn't this expect None max?

File: torch/_export/constraints.py [hunk 6]
```diff
@@ -1,48 +1,32 @@
...
-    # TODO: we should investigate turning off 0/1 specialization for unbacked
-    # SymInts
-    if min < 2:
-        raise ValueRangeError(
-            "Unable to set min size to be <= 2 because we specialize on 0/1 sizes."
-        )
-    return constrain_as_value(symbol, min, max)
+    # NOTE: If min, max value are not passed, we will assume it means this is only used for compiler hint.
+    # Runtime will assume min value will be 0 and max value will be INT_MAX
+    torch.sym_constrain_range(symbol, min=None, max=None)
```
Comment (ezyang) [inline torch/_export/constraints.py:32]: This doesn't seem like it will work? I feel like you need two ops.

File: aten/src/ATen/native/native_functions.yaml [hunk 7]
```diff
@@ -181,14 +181,21 @@
...
 - func: _assert_tensor_metadata(Tensor a, SymInt[]? size=None, SymInt[]? stride=None, ScalarType? dtype=None) -> ()
 
-- func: sym_constrain_range(Scalar size, int? min=None, int? max=None) -> ()
+- func: sym_constrain_range(Scalar size, *, int? min=None, int? max=None) -> ()
   dispatch:
-    CPU: sym_constrain_range_cpu
-    CUDA: sym_constrain_range_cuda
+    CompositeExplicitAutograd: sym_constrain_range
+
+- func: sym_constrain_range_for_size(Scalar size) -> ()
```
Comment (ezyang) [inline aten/src/ATen/native/native_functions.yaml:188]: This naming is weird because there's no range anymore. Why not just call it the public API name (`constrain_as_size`) (I'm also super unclear about why this operators have `sym_` prefix on their name, it seems unnecessary)

File: test/dynamo/test_export.py [hunk 8]
```diff
@@ -2321,7 +2321,7 @@ def test_exported_graph_serialization(self):
 
         def f(x, y):
             b = x.item()
-            constrain_as_size(b, min=2, max=5)
+            constrain_as_size(b)
             return torch.empty((b, y.shape[0]))
```
Comment (ezyang) [inline test/dynamo/test_export.py:2325]: 🤔  Maybe we should automatically constrain as size whenever you pass data dependent values in sizey contexts. This would be consistent with the "special irrefutable assert" construct we've been talking about.

Comment (ezyang) [reply test/dynamo/test_export.py:2325]: (no action needed)

File: torch/_custom_op/impl.py [hunk 9]
```diff
@@ -766,7 +766,7 @@ def create_unbacked_symint(self, *, min=2, max=None) -> torch.SymInt:
             )
 
         result = self._shape_env.create_unbacked_symint()
-        torch.fx.experimental.symbolic_shapes.constrain_range(result, min=2, max=max)
+        torch.fx.experimental.symbolic_shapes.constrain_range(result, min=2, max=max if max else sys.maxsize)
```
Comment (ezyang) [inline torch/_custom_op/impl.py:769]: Is this change required?

I have a good reason to not want to pass sys.maxsize as the upper bound: it will result in extra arbitrary precision integer math in many situations. For example, suppose I have x bounded (2, inf), if I multiple x by two its modified bound is (4, inf) (easy), but if I had maxsize, the modified bound is (4, maxsize * 2) (which is only representable in arbitrary precision.)

File: test/export/test_export.py [hunk 10]
```diff
@@ -37,23 +37,6 @@ def f(x):
...
-            export(invalid_size, inp)
-
-        def invalid_input_conflict_with_inline_constraints(x):
-            b = x.item()
-            constrain_as_size(b, min=2, max=5)
-            return torch.full((b, 1), 1)
-
-        inp = (torch.tensor([6]),)
-        with self.assertRaisesRegex(torchdynamo.exc.UserError, "Invalid value 6 for range"):
-            export(invalid_input_conflict_with_inline_constraints, inp)
```
Comment (ezyang) [inline test/export/test_export.py:56]: You can still do this test, so long as you actually run the resulting exported module (it should error at runtime)

File: test/export/test_export.py [hunk 11]
```diff
@@ -72,8 +55,8 @@ def invalid_input_conflict_with_input_constraints(x):
 
         def conflicting_constraints(x):
             b = x.item()
-            constrain_as_size(b, min=2, max=3)
-            constrain_as_size(b, min=4, max=5)
+            constrain_as_size(b)
+            constrain_as_value(b, min=4, max=5)
```
Comment (ezyang) [inline test/export/test_export.py:59]: This isn't really conflicting constraints anymore, right? Why is it erroring?

File: aten/src/ATen/native/Constraints.cpp [hunk 12]
```diff
@@ -20,16 +21,41 @@
...
 
-void sym_constrain_range_cpu(
+void sym_constrain_range(
     const Scalar& size,
     c10::optional<int64_t> min,
-    c10::optional<int64_t> max) {}
+    c10::optional<int64_t> max) {
 
-Tensor _functional_sym_constrain_range_cpu(
+    int64_t min_val = min.has_value() ? min.value() : 0;
```
Comment (ezyang) [inline aten/src/ATen/native/Constraints.cpp:29]: This is flat out wrong. If min is not specified, the minimum is unbounded below (could be negative). You need to simply skip the le check if it is none. Ditto below (where it is correct, but it will be more symmetric to skip the upper bound test when it's none.)

File: torch/_export/constraints.py [hunk 13]
```diff
@@ -1,48 +1,32 @@
...
     """
 
-    if not isinstance(symbol, SymInt):
-        constrain_range_int(symbol, min=min, max=max)
-    else:
-        sym_constrain_range(symbol, min, max)
+    if min is None:
+        min = 0
+    if max is None:
+        max = sys.maxsize
```
Comment (ezyang) [inline torch/_export/constraints.py:18]: I still don't think you should do this. What's the point of the op taking optional arguments if you always fill them in?

File: torch/fx/experimental/symbolic_shapes.py [hunk 14]
```diff
@@ -312,6 +312,39 @@ def _constrain_symbol_range(shape_env, s: sympy.Symbol, min: int, max: int):
     else:
         shape_env.var_to_range[s] = ValueRanges(min, max)
 
+    shape_env.runtime_var_to_range[s] = shape_env.var_to_range[s]
+
+def _constrain_unbacked_symint_for_size(a):
+    if isinstance(a, int):
+        return
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:319]: This is weirdly asymmetric with `_constrain_symbol_range`. Don't you want to assume you're getting a sympy.Symbol as argument?

Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:319]: (and then the enclosing code needs to do a >= 0 test if you do get an int)

File: torch/fx/experimental/symbolic_shapes.py [hunk 15]
```diff
@@ -312,6 +312,39 @@ def _constrain_symbol_range(shape_env, s: sympy.Symbol, min: int, max: int):
...
+
+    shape_env = a.node.shape_env
+    s = a.node.expr
+
+    if r := shape_env.var_to_range.get(s, None):
+        shape_env.var_to_range[s] = ValueRanges(
+            builtins.max(r.lower, 2), builtins.min(r.upper, sys.maxsize)
+        )
+    else:
+        shape_env.var_to_range[s] = ValueRanges(2, sys.maxsize)
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:335]: Don't use maxsize, as above

File: torch/_custom_op/impl.py (hunk 9 above)
Comment (tugsbayasgalan) [reply torch/_custom_op/impl.py:769]: Ohh this is not needed anymore. It was left from previous iteration.

File: torch/fx/experimental/symbolic_shapes.py [hunk 16]
```diff
@@ -312,6 +312,39 @@ def _constrain_symbol_range(shape_env, s: sympy.Symbol, min: int, max: int):
     else:
         shape_env.var_to_range[s] = ValueRanges(min, max)
 
+    shape_env.runtime_var_to_range[s] = shape_env.var_to_range[s]
```
[the remark this answers is not in the document]
Comment (tugsbayasgalan) [reply torch/fx/experimental/symbolic_shapes.py:315]: Hmm my thinking was if the same symbol that was used for `constrain_as_size` is used again to be constrained as value, we should respect the latter option.

[the remark this answers is not in the document]
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:315]: You do need to update runtime_var_to_range, but it shouldn't be an overwrite (an intersection, I think?)

[the remark this answers is not in the document]
Comment (tugsbayasgalan) [reply torch/fx/experimental/symbolic_shapes.py:315]: Yeah thats true.

File: test/test_proxy_tensor.py [hunk 17]
```diff
@@ -1128,7 +1130,7 @@ def f(x, mask, params, buffers):
                 for s in p.shape:
                     guard_int(s)
             x = x[mask]
-            constrain_range(x.shape[0], min=1)
+            constrain_as_value(x.shape[0], min=1)
```
Comment (ezyang) [inline test/test_proxy_tensor.py:1133]: Does `constrain_as_value` show up in the graph too? Hard to tell here.

File: aten/src/ATen/native/Constraints.cpp [hunk 18]
```diff
@@ -15,21 +16,48 @@
...
+Tensor _functional_sym_constrain_range(
     const Scalar& size,
     c10::optional<int64_t> min,
     c10::optional<int64_t> max,
     const Tensor& dep_token) {
+  sym_constrain_range(size, min, max);
+  return dep_token.clone();
+}
+
+void sym_constrain_for_size(const Scalar& size) {
```
Comment (ezyang) [inline aten/src/ATen/native/Constraints.cpp:56]: You need to test that it's an int, and that it's >= 0, no?

File: torch/_export/constraints.py [hunk 19]
```diff
@@ -1,48 +1,25 @@
...
-    # TODO: we should investigate turning off 0/1 specialization for unbacked
-    # SymInts
-    if min < 2:
-        raise ValueRangeError(
-            "Unable to set min size to be <= 2 because we specialize on 0/1 sizes."
-        )
-    return constrain_as_value(symbol, min, max)
+    # NOTE: If min, max value are not passed, we will assume it means this is only used for compiler hint.
+    # Runtime will assume min value will be 0 and max value will be INT_MAX
+    torch.sym_constrain_for_size(symbol)
```
Comment (ezyang) [inline torch/_export/constraints.py:25]: Why do we not line up the internal aten name with the public name here 🤔

File: torch/_export/constraints.py [hunk 20]
```diff
@@ -1,48 +1,25 @@
...
 
 # TODO: we want to hide this min/max stuff under some abstraction similar to
 # DynamicDim
-@allow_in_graph
-def constrain_as_size(symbol, min: int = 2, max: Optional[int] = None):
+def constrain_as_size(symbol):
     """
-    Add min/max constraint on the intermediate symbol which will be used as a size
+    Add compiler hints to the intermediate symbol which will be used as a size
+    in another tensor.
```
Comment (ezyang) [inline torch/_export/constraints.py:21]: Here's the docblock I would write:

This indicates that a given int is size-like, and can be used in any context where a size is expected. You will typically use this when reading out unbacked integers from Tensors, e.g., max.item() or lengths.tolist() which then need to be used as tensor constructors. Providing these assertions to PyTorch can help resolve GuardOnDataDependentSymNode errors upon export, since we cannot guard on unbacked SymInts.

This function has unusual semantics which distinguish it from constrain_as_value. Specifically, at compile-time, we will unsoundly assume that the resulting int is always `>= 2`. This makes it easier to use the unbacked int in size contexts, as we will often attempt to guard on a size being zero/one (e.g., when computing the contiguity of a tensor, or testing if broadcasting can occur), which will not work on unbacked SymInts. Assuming that the int is `>= 2` allows us to report False to these tests. Although this is technically unsound, in practice we observe that if your program works for all sizes `>= 2`, it probably works for zero and one too. At runtime, we only assert that the size-like int is `>= 0`, matching this intuition. (This is in contrast to `constrain_as_value(s0, min=2)` which would require, at runtime, that s0 is >= 2, or `constrain_as_value(s0, min=0)`, which typically isn't enough information to let you use s0 as a size to construct tensors.)

File: test/test_proxy_tensor.py (hunk 17 above)
Comment (tugsbayasgalan) [reply test/test_proxy_tensor.py:1132]: It will show up as aten.sym_constrain_range

File: torch/_export/constraints.py [hunk 21]
```diff
@@ -1,48 +1,59 @@
...
 
 # TODO: we want to hide this min/max stuff under some abstraction similar to
 # DynamicDim
-@allow_in_graph
-def constrain_as_size(symbol, min: int = 2, max: Optional[int] = None):
+def constrain_as_size(symbol, min: Optional[int] = None, max: Optional[int] = None):
     """
-    Add min/max constraint on the intermediate symbol which will be used as a size
+    This indicates that a given int is size-like, and can be used in any context where a size is expected.
+    You will typically use this when reading out unbacked integers from Tensors, e.g., max.item() or lengths.tolist()
```
Comment (gmagogsfm) [inline torch/_export/constraints.py:21]:
```suggestion
    You will typically use this when reading out integers from Tensors, e.g., max.item() or lengths.tolist()
```

Comment (gmagogsfm) [reply torch/_export/constraints.py:21]: users won't know what "unbacked" is

File: torch/_export/constraints.py [hunk 22]
```diff
@@ -1,48 +1,59 @@
...
+    which then need to be used as tensor constructors. Providing these assertions to PyTorch can help resolve
+      GuardOnDataDependentSymNode errors upon export, since we cannot guard on unbacked SymInts.
+
+    This function has unusual semantics which distinguish it from constrain_as_value.
+    Specifically, at compile-time, we will unsoundly assume that the resulting int is always >= 2.
```
Comment (gmagogsfm) [inline torch/_export/constraints.py:26]: Maybe mention the reason why we always assume it to be >= 2?

File: test/dynamo/test_export.py (hunk 8 above)
Comment (gmagogsfm) [reply test/dynamo/test_export.py:2326]: Yeah, not bad, we compromise a big of explicit-ness but users get less hassle in most cases.

File: torch/_export/__init__.py [hunk 23]
```diff
@@ -398,6 +398,8 @@ def convert_to_fake(x):
 
         except (ConstraintViolationError, ValueRangeError) as e:
             raise UserError(UserErrorType.CONSTRAIN_VIOLATION, str(e))
+        except torch._dynamo.exc.TorchRuntimeError as e:
```
Comment (gmagogsfm) [inline torch/_export/__init__.py:401]: All all runtime errors because of runtime assertion failures due to breaking constraints?

File: torch/_meta_registrations.py [hunk 24]
```diff
@@ -407,6 +412,19 @@ def functional_sym_constrain_range(size, min, max, dep_token):
...
+def sym_constrain_range_for_size(size, min=None, max=None):
+    if isinstance(size, (SymFloat, SymBool)):
+        raise ValueError("Constraining SymFloat or Symbool is nyi")
+    _constrain_range_for_size(size, min=min, max=max)
+
+
+@register_meta(aten._functional_sym_constrain_range_for_size.default)
+def functional_sym_constrain_range_for_size(size, min, max, dep_token):
+    aten.sym_constrain_range_for_size(size, min=min, max=max)
+    return dep_token
```
Comment (gmagogsfm) [inline torch/_meta_registrations.py:425]: Do we need to clone dep_token instead of returning itself directly?

File: torch/fx/experimental/symbolic_shapes.py [hunk 25]
```diff
@@ -304,13 +304,57 @@ def guard_scalar(a):
...
+            "received min={min} and max={max}"
+        )
+
+    compiler_min = 2 if min < 2 else min
+
+    _constrain_symbol_range(
+        a.node.shape_env,
+        a.node.expr,
+        compiler_min=compiler_min,
+        compiler_max=max,
```
Comment (gmagogsfm) [inline torch/fx/experimental/symbolic_shapes.py:353]: This `compiler_max` isn't really needed, are you doing it for symmetry?

File: test/dynamo/test_export.py [hunk 26]
```diff
@@ -2273,7 +2273,7 @@ def my_dyn_fn(a, b, c):
     def test_export_preserve_constraints_as_metadata_scalar(self):
         def f(x, y):
             b = x.item()
-            constrain_as_size(b, min=2, max=5)
+            constrain_as_size(b)
```
[the remark this answers is not in the document]
Comment (tugsbayasgalan) [reply test/dynamo/test_export.py:2276]: We do. I have some other tests that use min/max value. I was too lazy to change this back lol.

File: torch/_meta_registrations.py (hunk 24 above)
Comment (tugsbayasgalan) [reply torch/_meta_registrations.py:425]: I think that is only true for Cpp kernels.

File: torch/fx/experimental/symbolic_shapes.py (hunk 25 above)
Comment (tugsbayasgalan) [reply torch/fx/experimental/symbolic_shapes.py:353]: Yeah just for symmetry.

File: torch/_export/__init__.py (hunk 23 above)
Comment (tugsbayasgalan) [reply torch/_export/__init__.py:401]: I guess it is ok to directly raise TorchRuntimeError.

File: test/dynamo/test_export.py (hunk 8 above)
Comment (gmagogsfm) [reply test/dynamo/test_export.py:2326]: s/big/bit/

**Review by gmagogsfm — APPROVED** [review]

**Review by ezyang — APPROVED** [review]

File: torch/_meta_registrations.py (hunk 24 above)
Comment (ezyang) [reply torch/_meta_registrations.py:451]: No, you need to clone here too lol

**Comment by izaitsevfb** [comment]
@pytorchbot revert -m "Breaks inductor test on trunk" -c nosignal

see: https://github.com/pytorch/pytorch/actions/runs/5829257494/job/15808815449
